### ISSUE TITLE
Gemini 1.5 Flash: Candidate was blocked due to RECITATION when responseMimeType is json

### ISSUE THREAD

**Issue by marian2js** [google-gemini/generative-ai-js#138]
### Description of the bug:

When responseMimeType: 'application/json', a request is failing with error:
[GoogleGenerativeAI Error]: Candidate was blocked due to RECITATION.

However, without responseMimeType, the same prompt works (returns a markdown with json).

The exact same instructions and prompt work on the AI Studio, even with output in JSON on.

```typescript
// The error happens even if safety settings are set to block none.
const safetySettings = [
  {
    category: HarmCategory.HARM_CATEGORY_HARASSMENT,
    threshold: HarmBlockThreshold.BLOCK_NONE,
  },
  {
    category: HarmCategory.HARM_CATEGORY_HATE_SPEECH,
    threshold: HarmBlockThreshold.BLOCK_NONE,
  },
  {
    category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
    threshold: HarmBlockThreshold.BLOCK_NONE,
  },
  {
    category: HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,
    threshold: HarmBlockThreshold.BLOCK_NONE,
  },
]

const model = this.genAI.getGenerativeModel({
  model: 'gemini-1.5-flash-latest',
  systemInstruction: instructions,
  safetySettings,
})

const generationConfig = {
  temperature: 0,
  topP: 0.95,
  topK: 64,
  maxOutputTokens: 8192,
  responseMimeType: 'application/json', // fails only if this option is sent. 
}

const chatSession = model.startChat({
  generationConfig,
})

const result = await chatSession.sendMessage(prompt)
const text = result.response.text() // throws [GoogleGenerativeAI Error]: Candidate was blocked due to RECITATION.
```

### Actual vs expected behavior:

Actual: Throws [GoogleGenerativeAI Error]: Candidate was blocked due to RECITATION
Expected: Return the same result as in the AI Studio.

**Comment by ryanwilson** [comment]
Hi @marian2js, sorry for the troubles. Does *every* prompt cause this, or only specific prompts? If it's a specific prompt, are you able to share the prompt so we can try to reproduce it?

**Comment by marian2js** [comment]
Hi @ryanwilson. The issue happens only with a very specific prompt that I cannot share publicly. I've been trying to remove the personal data from it, but as soon as I do, it starts working.

I noticed that when I remove the responseMimeType, the JSON returned in the markdown is invalid as it has a js variable: `{ "key": value }`. However, with a different prompt the model returned the invalid json, so I don't know if the issue is related to that.

I am sorry for not being of more help.

**Comment by ryanwilson** [comment]
No worries! Out of curiousity, do you run into the same issue if you use `sendMessageStream` instead of `sendMessage`? That could be the difference with AI Studio, where the response is streamed.

**Comment by marian2js** [comment]
Hi @ryanwilson, the bug doesn't happen with `sendMessageStream`. I made changes to my prompt, so this bug is not triggered anymore for me. But I can confirm the bug is still happening with my old prompt and `sendMessage`.

**Comment by florian583** [comment]
Hi guys,

I'm experiencing the exact same problem as @marian2js. We're using Gemini with Vertex to extract structured data (as JSON) from job offer listing PDFs.

Here's what I've tried so far:

- Using the 1.5 Pro model instead of the Flash model
- Switching to `sendMessageStream` from `sendMessage`
- Using 1.0 Pro (works, but the output is non-usable)

Oddly enough, the issue seems to only happen when processing non-English documents. When I upload PDFs in French, I always run into the RECITATION problem.

On the flip side, if I use Google Vision for OCR on the PDF, and then use the Vercel AI SDK with chat and Gemini 1.5 Flash, it works perfectly with the same prompt, but on the OCR data (string) instead of the inline PDF.

Hope this info helps in figuring out the issue!

**Comment by GuyVivedus** [comment]
Hi folks,

Running into the same problem as @marian2js (with almost identical API call settings but with Golang - model conf, generationConfig, SafetyConfig all the same). Again parsing a plaintext file with a big prompt to output JSON (where the plaintext was originally converted from PDF). 

Setting model.GenerationConfig.ResponseMIMEType = "application/json" hits the "blocked: candidate: FinishReasonRecitation" resp with no PromptFeedback value, annoyingly, so flying a bit blind as to the cause. 

If it's helpful the content is senior school syllabus so shouldn't be anywhere close to hitting any of the harm response thresholds either way. 

Commenting ResponseMIMEType with a slight prompt modification gets a usable (yet obviously inconsistent) result. 

Much appreciated!

**Comment by chanmathew** [comment]
Experiencing same issue as well, same setup as others above.

**Comment by ShivQumis** [comment]
Changing to stream mode worked for me as well:

 ```
const safetySettings = [
    {
      category: HarmCategory.HARM_CATEGORY_HARASSMENT,
      threshold: HarmBlockThreshold.BLOCK_NONE,
    },
    {
      category: HarmCategory.HARM_CATEGORY_HATE_SPEECH,
      threshold: HarmBlockThreshold.BLOCK_NONE,
    },
    {
      category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
      threshold: HarmBlockThreshold.BLOCK_NONE,
    },
    {
      category: HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,
      threshold: HarmBlockThreshold.BLOCK_NONE,
    },
];


try {
            const chatSession = this.model.startChat({
                generationConfig: {
                    temperature: 1,
                    topP: 0.95,
                    topK: 64,
                    maxOutputTokens: 8192,
                    responseMimeType: "text/plain",
                },
                history: history,
                systemInstruction: systemMessage,
                safetySettings: safetySettings
            });
            
            var stream = true;
            if (stream) {
                const result = await chatSession.sendMessageStream(currentMessage);
              for await (const item of result.stream) {
                    console.log("Stream chunk: ", item.candidates[0].content.parts[0].text);
                } 
                const aggregatedResponse = await result.response;
                return aggregatedResponse.text();
            } else {
                const result = await chatSession.sendMessage(currentMessage);
                return result.response.text();
            }
        } catch (error) {
            console.error("Error during chat session:", error);
            throw error; // Re-throw the error after logging or handle it as needed
        }
```

**Comment by yharaskrik** [comment]
I started to receive this today as well, didn't get it at all yesterday and I made hundreds of calls, now today a significant number of them are receiving the error.

**Comment by johnomeara1** [comment]
Also getting this error.

**Comment by Kashi-Datum** [comment]
Any updates on this? As it stands at the current moment, using Gemini Pro 1.5 for RAG-base Q/A with chat history is incredibly unreliable.

Simply ask 2-3 questions about a document with some overlap, and you will encounter a ```RECITATION``` error almost guranteed.

I.e, using LangChain:

```
const chatTemplate = ChatPromptTemplate.fromMessages([['system', 'You are a helpful assistant. Answer all questions to the best of your ability.'], new MessagesPlaceholder('history'), new MessagesPlaceholder('input')]);
	const chain = chatTemplate.pipe(pdfModel).pipe(new StringOutputParser());

const chainWithHistory = new RunnableWithMessageHistory({
		runnable: chain,
		getMessageHistory: () => chatHistory,
		inputMessagesKey: 'input',
		historyMessagesKey: 'history',
		config: { configurable: { sessionId: decodedToken.uid } }
	});

	const response = await chainWithHistory.invoke({ input: input });
```

If you simply ask the model to recite a specific section of a file/PDF twice in a row, you will get a Recitation error GURANTEED (whether you pass the input File in once at the beginning of the conversation, only in the latest message, or as part of every message input is irrelevant).

**Comment by Aftab-M** [comment]
I'm getting the same error. 
Isn't there a way to know in advance if the input contains something which might cause this error, and then skip that ??

**Comment by HarshavardhanNetha** [comment]
> Hi @ryanwilson, the bug doesn't happen with `sendMessageStream`. I made changes to my prompt, so this bug is not triggered anymore for me. But I can confirm the bug is still happening with my old prompt and `sendMessage`.

I have tried the same, it seems the issue still persists with sendMessageStream as well.
One more interesting thing I found was, when using sendMessageStream, along with some tools, it is producing 500 Internal Server Error.

**Comment by 0xthierry** [comment]
I have the same problem, I'm sending the screenshot of a PDF, it works in some cases but not in others

**Comment by jouwana** [comment]
i also have the same problem.

i ran 1.0 pro vs 1,5 flash (as i got the email for it becoming deprecated soon) - i tried 1.5 pro as well

when running 1.0 & 1.5 pro, i have zero issues, no matter how many time i retry the same content generation in a row, or how many times i send (assuming im not limited)

![pro](https://github.com/google-gemini/generative-ai-js/assets/41836591/3b9cab99-6fce-4c50-8ae5-9c4a5e119f6c)

yet, when i run gemini-1.5-flash, exact same prompt, exact same code, i would randomly get the recitation error (that i am currently not handling in catch to just see it easily on console)

![flash](https://github.com/google-gemini/generative-ai-js/assets/41836591/60702436-dcdc-4cd6-9079-3719901d6a1d)

thing is, its not even particularly faster than the 1.0 pro. the 1.5 pro also does not give this error.

i am running the free version so im comparing 1.0 pro and 1,5 flash mostly as they both have similar request limitations.

**Comment by HarshavardhanNetha** [comment]
> i also have the same problem.
> 
> i ran 1.0 pro vs 1,5 flash (as i got the email for it becoming deprecated soon) - i tried 1.5 pro as well
> 
> when running 1.0 & 1.5 pro, i have zero issues, no matter how many time i retry the same content generation in a row, or how many times i send (assuming im not limited)
> 
> ![pro](https://private-user-images.githubusercontent.com/41836591/339198417-3b9cab99-6fce-4c50-8ae5-9c4a5e119f6c.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNTA0NjUsIm5iZiI6MTcxODI1MDE2NSwicGF0aCI6Ii80MTgzNjU5MS8zMzkxOTg0MTctM2I5Y2FiOTktNmZjZS00YzUwLThhZTUtOWM0YTVlMTE5ZjZjLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDAzNDI0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0M2M4N2Y1YjAwMzczZDEyMmEwNzQ2OWFiNDAzMzQ2ZjA3NmM0YjcyMmE1MWRjNTkxNjU4NWM2MTA3YTJhZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hqWGewmsObFcuX8S-KgVhO170Q0NzGJDht7ZWP3gM_0)
> 
> yet, when i run gemini-1.5-flash, exact same prompt, exact same code, i would randomly get the recitation error (that i am currently not handling in catch to just see it easily on console)
> 
> ![flash](https://private-user-images.githubusercontent.com/41836591/339200790-60702436-dcdc-4cd6-9079-3719901d6a1d.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNTA0NjUsIm5iZiI6MTcxODI1MDE2NSwicGF0aCI6Ii80MTgzNjU5MS8zMzkyMDA3OTAtNjA3MDI0MzYtZGNkYy00Y2Q2LTkwNzktMzcxOTkwMWQ2YTFkLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDAzNDI0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhYmZiNjFkOTU3ZWYxMmM1MGM3MWE2MWI5NzdhNjEzZmU1YzVmNGYwNWFiZDE5ZWQ2OGQ5MzdjOThmMzcyMWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0Zx3uoGbsfCCZoczfC26peAxR7PbVJ1CDQ9xYq_ka_o)
> 
> thing is, its not even particularly faster than the 1.0 pro. the 1.5 pro also does not give this error.
> 
> i am running the free version so im comparing 1.0 pro and 1,5 flash mostly as they both have similar request limitations.
> 
> i did not change any settings, didnt set up safety or response type or anything, so all there should be the default

By any chance, have you tried using tools? And experimented with various models?
Also, any work done on sendMessage & sendMessageStream?

**Comment by jouwana** [comment]
> 
> By any chance, have you tried using tools? And experimented with various models? Also, any work done on sendMessage & sendMessageStream?

i am unsure what 'tools' is, so probably havent tried it.

i have tried different things with the different models, in prompt form, length, sending same one in a row vs sending different ones in a row, only Flash has the problem, and if pops up no matter what prompts im sending, usually after 2-3 sends, sometimes ca last to 4-5 without error but not as common.

as for sendMessage and sendMessageStream, i am mostly sending unrelated prompts, which is why i preferred using generateContent over openChat and sendMessage, and so i havent tried them for now.

**Comment by chanmathew** [comment]
@hsubox76 just wanted to page contributors here, this is a MAJOR bug, especially when you guys are sunsetting 1.0 pro and moving everyone to 1.5 flash. It is literally unusable right now with this recitation error. This is the only reason we cannot migrate over from other foundation models. You are losing out on customers. 

Is anyone on the team looking at this?

**Comment by httplups** [comment]
Same problem here using Gemini 1.5 Pro

**Comment by yeiinissi** [comment]
I have this problem using gemini-1.5-flash. If I send the same prompt a second time it shows me this error:
 [GoogleGenerativeAI Error]: Candidate was blocked due to SAFETY
 
 Prompt example: 
 Write a romantic sentence that includes the number '5' written exactly as '5'. Make sure the number appears in numerical format and not in words.
 
 If you send the same thing but changing only the number, it generates the error. For example:
Write a romantic sentence that includes the number '73' written exactly as '73'. Make sure the number appears in numerical format and not in words.

**Comment by rartin** [comment]
Same issue! Looks like this issue is stock under the backlog pile.

**Comment by hsubox76** [comment]
This does seem like a serious issue - unfortunately it's beyond the ability of us to fix in the SDK. We're asking anyone to bring issues with the service or the models to the discussion forum: https://discuss.ai.google.dev/ where it will be more likely to reach those who cover the models themselves.

If posting in the forum, feel free to link back to this issue as it contains a lot of info and examples of the problem which might be helpful background, and if someone opens a thread in the discussion forum, please link it here so that all the JS SDK users facing this issue can add to that thread and hopefully get it looked at.

We will also try to get some answers internally but it's probably best for us to use all channels, including the discussion forum.

**Comment by logankilpatrick** [comment]
Hey folks, following up, we are investigating this and will share more details as soon as we have an update on our end!

**Comment by matadornetwork** [comment]
> Hey folks, following up, we are investigating this and will share more details as soon as we have an update on our end!

Hi Logan, any updates on this?

**Comment by pinballsurgeon** [comment]
fyi - Here is a recitation failure I find reproducible and strange, the behavior/error differences between nodejs and console may be telling, perhaps the listing/encoding causes sensitivity but no changes to parameters can prevent error - 

**NO ERROR** - CHESS OPENINGS - instructions prompt - 
System: You are a service that returns tidy, single line list of domain items responses to prompts. Your response is only a comma seperated lists of items like 'item1, item2, item3..'. List as many items as performantly possible to enrich the data. If prompted for a list of fruits, you respond with 'apple, pear, banana, ..'. Do not include ‘..’ in the response, instead add as many items (10-50) as needed to fulfil the request in a timely manner. \n\nHuman: Given these instructions, the prompt asks for a comma separated list of **chess openings**, return your tidy response of items. \n\nAssistant: Here is the requested list of comma seperated domain items; 

**RECITATION ERROR** - POKEMON - instructions prompt - 
System: You are a service that returns tidy, single line list of domain items responses to prompts. Your response is only a comma seperated lists of items like 'item1, item2, item3..'. List as many items as performantly possible to enrich the data. If prompted for a list of fruits, you respond with 'apple, pear, banana, ..'. Do not include ‘..’ in the response, instead add as many items (10-50) as needed to fulfil the request in a timely manner. \n\nHuman: Given these instructions, the prompt asks for a comma separated list of **pokemon**, return your tidy response of items. \n\nAssistant: Here is the requested list of comma seperated domain items;

**Comment by saranggupta94** [comment]
I am facing the same issue. Following this thread.

**Comment by shashank734** [comment]
Facing the same issue with 1.5 Pro.

**Comment by Samson-DB** [comment]
Facing the same issue for a RAG use case using gemini-1.5.pro. It's very unpredictable as the error depends on the output and not always the context/prompt.

**Comment by anthonylee991** [comment]
Ran into this issue as well today.

**Comment by iliane5** [comment]
Having the same issues with flash and pro. Impossible to use them in production.

**Comment by stebansaa** [comment]
having this same problem

**Comment by MatejaTrik** [comment]
Same issue here, although i am summarizing books from PDF provided by me. Could it be something about Authors' rights?

**Comment by jamesjhonatan123** [comment]
same issue here

**Comment by matadornetwork** [comment]
> Hey folks, following up, we are investigating this and will share more details as soon as we have an update on our end!

Logan any update?

**Comment by 0xSMW** [comment]
can confirm same issue still

**Comment by logankilpatrick** [comment]
Hey folks, I am going to declare an internal code red on this until it is resolved, sorry it has taken so long. Will provide updates twice a week until it is resolved.

**Comment by logankilpatrick** [comment]
Update: We have found what we believe to be part of the problem when sending requests with JSON mode enabled. Potential fix in the works, will share more on the timeline once we see how it goes.

Note that it's not clear everyone in this thread has the same issue, the current fix we are looking at is for specifically requests sent with JSON mode enabled using 1.5 Flash and 1.5 Pro.

**Comment by chanmathew** [comment]
> Update: We have found what we believe to be part of the problem when sending requests with JSON mode enabled. Potential fix in the works, will share more on the timeline once we see how it goes.
> 
> Note that it's not clear everyone in this thread has the same issue, the current fix we are looking at is for specifically requests sent with JSON mode enabled using 1.5 Flash and 1.5 Pro.

You're the best Logan! Appreciate the urgency! 👏

**Comment by ShivQumis** [comment]
> Update: We have found what we believe to be part of the problem when sending requests with JSON mode enabled. Potential fix in the works, will share more on the timeline once we see how it goes.
> 
> Note that it's not clear everyone in this thread has the same issue, the current fix we are looking at is for specifically requests sent with JSON mode enabled using 1.5 Flash and 1.5 Pro.

Just pointing out significant amount of examples with errors do not have JSON mode enabled.

**Comment by matadornetwork** [comment]
> Just pointing out significant amount of examples with errors do not have JSON mode enabled.

Yes echoing this statement. We found it happening with both JSON mode enabled and disabled.

**Comment by baxteran** [comment]
Hi, I'm also suffering from the RECITATION problem when using Gemini Pro 1.5 with llamaIndex CondensePlusContextChatEngine. The problem only occurs (for me) on subsequent turns of the chat, not the first. Whether an exception is thrown seems to be dependent on the RAG and the user question, but is repeatable for the same conversation sequences. Looking forward to your fix - thanks very much!

**Comment by logankilpatrick** [comment]
Update: we discovered a bug related to JSON Schema mode and recitation checks, we rolled out the change yesterday and are monitoring the impact. This should help the issue of recitation errors when in JSON mode. 

Note that as mentioned, this doesn't capture all of the recitation comments in this thread which is a mix of different use cases. We have not identified any known issues with other parts of the recitation checking at this time but are still investigating.

**Comment by iliane5** [comment]
> We have not identified any known issues with other parts of the recitation checking at this time but are still investigating.

If sending problematic prompts helps troubleshooting, I think everyone in this thread and on the bug tracker (https://issuetracker.google.com/issues/331677495?pli=1) would be more than happy to send a few ;)

> The inconsistent but frequent occurrence (almost daily) of this error effectively makes the API useless for our use-case and we switched back to Claude.

Same here. I'm still hoping for the bug to be fixed at some point but it looks like this is somehow expected behaviour (?) if no issue with the recitation checker has been found.

**Comment by gitcagey** [comment]
Thanks @logankilpatrick! We have been fighting with this over past week trying to pull glossary terms out some documents we use for work. Almost to point where was going to have to consider changing LLMs. 

I changed some of our prompting around today and did not get the error where we usually do (json mode enabled using 1.5 pro). Thought was my prompting but maybe was this change. Will test further.

**Comment by KlimentP** [comment]
Can confirm that previous test examples (gemini flash 1.5 in JSON mode) that used to cause the error, no longer cause issues

**Comment by benripka** [comment]
Our users get hundreds of RECITATION errors a day, with JSON mode **disabled**. Luckily we fallback to other LLM APIs who do not suffer from such a glaring flaw, and can handle the exact same prompts fine. Typically, the error occurs when the prompt requests that the LLM repeat something from earlier in the conversation, with some formatting change, for example. Many such prompts result in RECITATION errors. Is this the intent behind the RECITATION blocker? That it prevents the model from repeating text strings from earlier in the chat conversation? If this is the case than it greatly reduces the utility of the API. I was under the impression that the intent was to block reciation of **training** data.

**Comment by gitcagey** [comment]
Any update on this? it is back to blocking us now. We are trying to use gemini to extract key terms/glossary from documents. There are some docs that it fails nearly 100% of time with, even with some of the creative prompting.

**Comment by asaxena1415** [comment]
hi, is there any update on this?

**Comment by ronneldavis** [comment]
I faced this issue now with JSON mode disabled. The use case is code generation which seems pretty basic.

**Comment by rodrisapiens** [comment]
This model cannot be used with RAG nor perform OCR on book pages because of the recitation restrictions.

**Comment by DEVBALIYAN21** [comment]
YOU GUYS CAN TRY THIS CODE IT WORKED FOR ME AND GENERATING THE RESULTS 
I HAVE USE NPM



import { GoogleGenerativeAI, HarmCategory, HarmBlockThreshold } from "@google/generative-ai";

// Fetch your API_KEY
const API_KEY = "";  // Replace with your actual API key

// Initialize the AI model
const genAI = new GoogleGenerativeAI(API_KEY);
const model = genAI.getGenerativeModel({ model: "gemini-1.5-flash" });

// Define safety settings
const safetySettings = [
    {
        category: HarmCategory.HARM_CATEGORY_HARASSMENT,
        threshold: HarmBlockThreshold.BLOCK_NONE,
    },
    {
        category: HarmCategory.HARM_CATEGORY_HATE_SPEECH,
        threshold: HarmBlockThreshold.BLOCK_NONE,
    },
    {
        category: HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT,
        threshold: HarmBlockThreshold.BLOCK_NONE,
    },
    {
        category: HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT,
        threshold: HarmBlockThreshold.BLOCK_NONE,
    },
];

document.getElementById("generate").addEventListener("click", async () => {
    const prompt = document.getElementById("prompt").value;
    const codeDiv = document.getElementById("code");
    const contentDiv = document.getElementById("content");
    codeDiv.innerHTML = "";
    contentDiv.innerHTML = "Generating...";

    try {
        const chatSession = model.startChat({
            generationConfig: {
                temperature: 1,
                topP: 0.95,
                topK: 64,
                maxOutputTokens: 8192,
                responseMimeType: "text/plain",
            },
            safetySettings: safetySettings
        });

        var stream = true;
        let resultText = '';

        if (stream) {
            const result = await chatSession.sendMessageStream(prompt);
            
            // Process the streaming response
            for await (const chunk of result.stream) {
                if (chunk.candidates && chunk.candidates[0].content && chunk.candidates[0].content.parts) {
                    resultText += chunk.candidates[0].content.parts.map(part => part.text).join('');
                }
            }
            
            // Finalize the result
            contentDiv.innerHTML = `<pre>${resultText}</pre>`;
        } else {
            const result = await chatSession.sendMessage(prompt);
            resultText = await result.response.text();
            contentDiv.innerHTML = `<p>${resultText}</p>`;
        }
    } catch (error) {
        console.error("Error generating content:", error);
        contentDiv.innerHTML = handleErrors(error);
    }
});

function handleErrors(error) {
    if (error.message.includes("RECITATION") || error.message.includes("content blocked")) {
        return "The generated content was blocked. Please try a different prompt or adjust your input.";
    } else {
        return `An error occurred: ${error.message}. Please try again later.`;
    }
}

**Comment by logankilpatrick** [comment]
We just shipped another update which should help reduce recitation blocks in cases where the context was sent in the request (RAG situations).

**Comment by rodrisapiens** [comment]
> We just shipped another update which should help reduce recitation blocks in cases where the context was sent in the request (RAG situations). 

Are you planning to do the same with OCR implementations? 

I'm giving Gemini flash a book's page image for it to return the text in it and it's giving me the recitation error making it unusable.

It doesn't make any sense since this book was written by Tolstoy (War and peace), no copyright infringement whatsoever.

**Comment by logankilpatrick** [comment]
I will look into this ^ 

We also have another change in the works that should decrease the chance you get a recitation block on a simple code sample, it will take about ~2 weeks to fully roll this out, but stay tuned!

**Comment by meheli-basu** [comment]
> We just shipped another update which should help reduce recitation blocks in cases where the context was sent in the request (RAG situations).

I am trying gemini-1.5-pro-001 with pdf's and have used all sort of block levels, including BLOCK_NONE for the safety settings. And every time run into the error: he candidate is likely blocked by the safety filters......  "finish_reason": "RECITATION",
<img width="535" alt="erroe" src="https://github.com/user-attachments/assets/d82f997e-3bb3-4042-be85-efeba1150376">

Is this error particularly bad with pdf's or does it not matter?

**Comment by jsomeara** [comment]
This issue is the only thing preventing me from using Gemini. Very sad...

**Comment by JacobAsmuth** [comment]
Hey @logankilpatrick! Has this change rolled out yet?

**Comment by rodrisapiens** [comment]
fyi, i tried streaming too for OCR, but it stops due to RECITATION. Are you planning to solve this ?

**Comment by JamMaster1999** [comment]
I have the same issue with OCR @logankilpatrick. I am providing with public research papers and it suddenly stops and says 
![image](https://github.com/user-attachments/assets/4d4e4cdc-a1c3-4785-acc6-221c1b6a1696)
Really disappointing to see a great model become absolutely unusable on so many use cases.

**Comment by bn-l** [comment]
Just chiming to add that for almost all OCR tasks I am getting this error.

**Comment by bcfoltz** [comment]
Trying to condense some PDF content down into manageable markdown summaries and this RECITATION error is just a complete dealbreaker using Flash. Random, annoying, and silly. Temperature set to 1. Pro seems to work better but still throws the RECITATION error, just a little less often.

Going to have to use PyPDF probably and switch to a different product. Unusable for my business.

**Comment by jsomeara** [comment]
Google doesn't seem to realize just how many people this one bug is turning
away.

**Comment by heymattiarighetti** [comment]
Literally. I wanted to switch from Claude 3.5 Sonnet to Gemini 1.5 Pro in my app, but I couldn't run the test because 9 times out of 10 returns an error between RECITATION or SAFETY (for things that are totally safe 😅)

Not realiable = immediate turn off

**Comment by gsweeh** [comment]
I’m using the Gemini 1.5 Pro EXP model API with the image shown below, but I’m encountering the same issue. Here’s the error I’m receiving:

**Response:**

```
python
GenerateContentResponse(
    done=True,
    iterator=None,
    result=protos.GenerateContentResponse({
      "candidates": [
        {
          "finish_reason": "RECITATION",
          "index": 0
        }
      ],
      "usage_metadata": {
        "prompt_token_count": 407,
        "total_token_count": 407
      }
    }),
)
```

**Here’s the code I’m using:**

```
import google.generativeai as genai
import os
import PIL.Image
from dotenv import load_dotenv
from google.generativeai.types import HarmCategory, HarmBlockThreshold

load_dotenv()

# Configure the API key
genai.configure(api_key=os.environ["GEMINI_API_KEY4"])

# Define generation configuration
generation_config = {
    "temperature": 0,
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

# Create a GenerativeModel instance
model = genai.GenerativeModel(
    model_name="gemini-1.5-pro-exp-0827",
    generation_config=generation_config,
    safety_settings={
        HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,
        HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,
        HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,
        HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,
    },
    system_instruction="""
    Extract questions and their corresponding options (if applicable) from an image of a question paper, including any directions, instructions, and question numbers. The extracted content should be formatted in LaTeX code, preserving the original structure and layout of the question paper. 

    - **Questions**: Format each question with its number.
    - **Options**: Use the enumerate environment for multiple-choice options, labeling each option as (a), (b), (c), etc.
    - **Diagrams**: If a question includes a diagram, use the TikZ package to recreate the diagram in LaTeX. The TikZ code should be appropriately placed within the LaTeX document, ensuring the diagram aligns with the relevant question.
    """
)

# Path to the images folder
images_folder = "images"
image_files = [f for f in os.listdir(images_folder) if f.endswith('.png')]

# Process each image file
for image_file in image_files:
    image_path = os.path.join(images_folder, image_file)
    sample_image = PIL.Image.open(image_path)

    # Generate content based on the image
    response = model.generate_content([sample_image])

    # Print the generated LaTeX code or handle the response as needed
    print(response)
```

**When I use the same image and system instruction in the Gemini AI Studio Playground, I receive the correct response. You can see the following images:**

1.The first image shows the API response.
![Screenshot 2024-09-03 123604](https://github.com/user-attachments/assets/0721c6ee-92d6-4f6a-af30-dc7d9dc5f520)

2. The second image shows the Gemini AI Studio Playground response.
![Screenshot 2024-09-03 123426](https://github.com/user-attachments/assets/08c17905-1da9-4b7a-94f9-956aba1d09d9)

3. The third image shows the safety settings.
![Screenshot 2024-09-03 123700](https://github.com/user-attachments/assets/542592d8-0282-473a-9b5b-3d6483c97ff0)

4. The fourth image is the sample image to process.
![question1 (12)](https://github.com/user-attachments/assets/cbca4202-fa4f-467c-ba90-cbf1a3226493)

**Comment by altbdoor** [comment]
Just re-echoing my sentiments in https://github.com/google/generative-ai-docs/issues/257#issuecomment-2327587546

Asking `what are the 12 star signs?` also triggers recitation error. At this rate, nearly all general knowledge in Gemini's data set would have triggered recitation error.

**Comment by ezanalemma** [comment]
Using the relatively new **`responseSchema`** parameter for JSON in the API seems to trigger RECITATION errors... never had them before.

**Comment by 3choff** [comment]
It seems the recitation problem has been fixed with the latest models. In my tests, answer generation no longer stops prematurely. Before I resume a project on hold due to this issue, I'm gathering feedback from other users about their experiences with this and the new models. Does anyone still see 'finish_reason: RECITATION' in those triggering requests when switching to the new models?

**Comment by ronneldavis** [comment]
Yeah it seems to have stopped with the latest models now, not tested it comprehensively but it seems better now.

**Comment by iliane5** [comment]
> It seems the recitation problem has been fixed with the latest models. In my tests, answer generation no longer stops prematurely. Before I resume a project on hold due to this issue, I'm gathering feedback from other users about their experiences with this and the new models. Does anyone still see 'finish_reason: RECITATION' in those triggering requests when switching to the new models?

They've improved it but the recitation filter is still fundamentally broken and will randomly block requests in an unpredictable manner. Just ran into another case a minute ago while doing some testing.

**Comment by naourass** [comment]
After doing some tests, it doesn't seem that the recitation has been fixed. I'm still getting ton of recitation errors when trying to extract data from provided input. All the requested data to extract is already in the input (law texts and attachments), so nothing sensitive or related to copyright.

**Comment by paul-at-modgenai** [comment]
Here's an example I'm seeing in my own enviroment:

```
    let result = await chatSession.sendMessageStream("Respond in  Markdown. Give me Tic Tac Toe in java.");
```

Sometimes it works sometimes it does not:

```
Error: [GoogleGenerativeAI Error]: Candidate was blocked due to RECITATION\n    at response.text (file:///D:/mais/mais-svc-gemini/node_modules/@google/generative-ai/dist/index.mjs:454:23)\n    at <anonymous> (D:\\mais\\mais-svc-gemini\\src\\index.ts:73:31)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
```

**Comment by sergeyk** [comment]
Not acceptable.

**Comment by simonw** [comment]
I've been hitting this too. I'm trying to feed in an image of a page of a PDF and then running this prompt against it:

> `Convert this image into markdown. If there are multiple columns on the page turn them into a linear list. Keep any references and put them in a <span class=reference> tag. For images and charts and figures return a <div name="title of figure" data-bbox="ymin, xmin, ymax, xmax">alt text description of that image</div> where those are the bounding box coordinates, and place that div in the relevant location between the paragraphs of text.`

**Comment by chippernut** [comment]
Having this issue too. Using Gemini Pro 1.5 for OCR via JSON. I found a temporary work-around by adjusting the image. My use case is the user manually snips a part of a larger image/PDF for upload to Gemini, the workaround is to take a new snip image but intentionally make it larger than the target to trick the AI into thinking this is a totally new image. The trick only works for one response, then the RECITATION error occurs again.  It's also not 100%, but did resolve some instances.

**Comment by nabildroid** [comment]
i proved a pdf and asked to extract the data, i see the RECITATION error :(

**Comment by hegdeadithyak** [comment]
I am also getting this error, any found some fix?

**Comment by kym6464** [comment]
I've also run into a similar refusal issue with JSON mode and created a reproducible example here https://github.com/kym6464/gemini-spii-sample

**Comment by altbdoor** [comment]
> `category: HarmCategory.HARM_CATEGORY_UNSPECIFIED,`

@vedtam `HARM_CATEGORY_UNSPECIFIED` is a _returned harm category from the API_. You should not send it in the API call.

The error message you had, can be roughly translated as:

> One of your `category` in `safetySettings`, is not a valid category. The valid categories are:
> - HARM_CATEGORY_HATE_SPEECH
> - HARM_CATEGORY_SEXUALLY_EXPLICIT
> - HARM_CATEGORY_DANGEROUS_CONTENT
> - HARM_CATEGORY_HARASSMENT
> - HARM_CATEGORY_CIVIC_INTEGRITY

**Comment by vedtam** [comment]
Thanks @altbdoor, that worked! Still, such a unintuitive error, they could just as easily print the bad category's name. The source code clearly lists UNSPECIFIED as a valid category:

```
/**
 * Harm categories that would cause prompts or candidates to be blocked.
 * @public
 */
export declare enum HarmCategory {
    HARM_CATEGORY_UNSPECIFIED = "HARM_CATEGORY_UNSPECIFIED",
    HARM_CATEGORY_HATE_SPEECH = "HARM_CATEGORY_HATE_SPEECH",
    HARM_CATEGORY_SEXUALLY_EXPLICIT = "HARM_CATEGORY_SEXUALLY_EXPLICIT",
    HARM_CATEGORY_HARASSMENT = "HARM_CATEGORY_HARASSMENT",
    HARM_CATEGORY_DANGEROUS_CONTENT = "HARM_CATEGORY_DANGEROUS_CONTENT"
}
```

**Comment by avelops** [comment]
i'm getting this too in cline (prev. Claude-Dev).

**Comment by ubuntupunk** [comment]
Asking Cline to place example json data for a Google Service account json file elicited this error on Gemini Pro.

**Comment by naourass** [comment]
> Hi, I had this same problem, almost half of my request were returning the 'RECITATION' finish reason. I noticed I was using the model "gemini-1.5-flash", and just switching to the model "gemini-1.5-flash-002" seems to have solved the problem.

Just tested both flash and pro `002` versions, unfortunately still getting tons of recitation errors when prompting to extract data from my own input.

**Comment by kevinrss01** [comment]
Same issue here :(

**Comment by h2210316651** [comment]
I was baffled to see citations of parts of the text that were blocked due to the recitation reason work perfectly fine in the gemini app with the same sources cited. What gives google the authority to recite someone else's work while people like us can't?? If you cite and say I got the text from this site, please extend us the same courtesy where you give us the text and let us do the citations.

**Comment by rgex** [comment]
Try adding the following in the prompt:

You are allowed to quote text from the document in order to provide the information requested but only if the quote is less than 8 words.

**Comment by XInTheDark** [comment]
The problem is still present in the latest experimental models.

**Comment by ElYaiko** [comment]
I have this problem with `gemini-exp-1206` but not `gemini-2.0-flash-exp`

**Comment by BlazingTide** [comment]
any updates on this? It's causing a lot of problems for me.

**Comment by BlazingTide** [comment]
the api is practically unusable atm.

**Comment by Icksir** [comment]
I hope this gets solved soon, it is incredibly unreliable even for creating summaries to some documents.

**Comment by wikisteff** [comment]
I had the blocked due to RECITATION problem today with a prompt where I was asking for examples to some distilled content on Gemini 2.0 Flash. Did not use responseMimeType. I overcame it by adding the following clause to the prompt:
"You are allowed to quote text from the document in order to provide the information requested but only if the quote is less than 50 words."

I didn't have a document, but it still made the problem go away.

**Comment by miguelsalcedo01** [comment]
This isn't just a text completion issue but also happening in VS code when using in Roo code, Cline, etc..  It's been causeing me a huge problem. Almost useless now, and I even make use of memory banks to keep my ide gemini agent aligned with the project i'm working on and now all of a sudden i'm seeing this much more often and 503 retry server too busy..

**Comment by stri8ed** [comment]
FYI, there is a new public bug on this topic in a more central place: https://issuetracker.google.com/issues/331677495

**Comment by logankilpatrick** [comment]
Hey folks, quick update, over the last few months, we have driven down recitation blocks by >75% without compromising on our obligations across the ecosystem. Pls keep the feedback coming if any issues come up!

**Comment by logankilpatrick** [comment]
I am going to close this for now since this issue has been open for a while, but thank you to everyone who has sent feedback about this. We have more robust tooling, teams in place, etc to handle issues like this in the future so pls keep the feedback coming!
